File: Cora/Objective-C/Controllers/UPTRootVC.m
#import "UPTRootVC.h"


@implementation UPTRootVC


- (id)initWithNibName:(NSString *)aNib bundle:(NSBundle *)aBundle {

	self = [super initWithNibName:aNib bundle:aBundle];

	if(self) {

		[self setupDefaults];
		[self setupUI];
		[self setupObservers];

		[NSTimer scheduledTimerWithTimeInterval:5 target:self selector:@selector(updateUptimeText) userInfo:nil repeats:YES];

	}

	return self;

}


- (void)setupDefaults {

	[[UserDefaultsManager sharedInstance] loadAccentColor];
	[[UserDefaultsManager sharedInstance] loadSwitchState];

	BOOL state = [UserDefaultsManager sharedInstance].switchState;

	if(state) [self setDarwinText];

}


- (void)setupObservers {

	[NSNotificationCenter.defaultCenter removeObserver:self];
	[NSNotificationCenter.defaultCenter addObserver:self selector:@selector(setDarwinText) name:@"launchChosenTask" object:nil];
	[NSNotificationCenter.defaultCenter addObserver:self selector:@selector(updateAccentColors) name:@"updateAccentColors" object:nil];

}


- (void)viewDidLoad {

	[super viewDidLoad];

	// Do any additional setup after loading the view, typically from a nib.

	self.view.backgroundColor = UIColor.systemBackgroundColor;

}


- (void)viewDidLayoutSubviews {

	[super viewDidLayoutSubviews];

	[self layoutUI];

}


- (void)setupUI {

	self.darwinLabel = [UILabel new];
	self.darwinLabel.font = [UIFont fontWithName:@"Courier" size:12];
	self.darwinLabel.text = [TaskManager sharedInstance].outputString;
	self.darwinLabel.textColor = [ColorManager sharedInstance].accentColor;
	self.darwinLabel.numberOfLines = 0;
	self.darwinLabel.textAlignment = NSTextAlignmentCenter;
	self.darwinLabel.translatesAutoresizingMaskIntoConstraints = NO;
	[self.view addSubview:self.darwinLabel];

	self.uptimeLabel = [UILabel new];
	self.uptimeLabel.font = [UIFont fontWithName:@"Courier" size:16];
	self.uptimeLabel.text = [self setUptimeText];
	self.uptimeLabel.textColor = [ColorManager sharedInstance].accentColor;
	self.uptimeLabel.numberOfLines = 0;
	self.uptimeLabel.textAlignment = NSTextAlignmentCenter;
	self.uptimeLabel.translatesAutoresizingMaskIntoConstraints = NO;
	[self.view addSubview:self.uptimeLabel];

	dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_LOW, 0), ^{

		[self animateLabel:[TaskManager sharedInstance].outputString withCharacterDelay:0.008];

	});

	self.settingsButton = [UIButton new];
	self.settingsButton.tintColor = [ColorManager sharedInstance].accentColor;
	[self.settingsButton setImage : [UIImage systemImageNamed:@"gear"] forState: UIControlStateNormal];
	[self.settingsButton addTarget : self action:@selector(didTapSettingsButton) forControlEvents: UIControlEventTouchUpInside];

	UIBarButtonItem *settingsButtonItem = [[UIBarButtonItem alloc] initWithCustomView: self.settingsButton];
	self.navigationItem.rightBarButtonItem = settingsButtonItem;

	[self updateAccentColors];

}


- (void)layoutUI {

	int safeInsetBottom = self.view.safeAreaInsets.bottom;

	NSDictionary *views = @{

		@"darwinLabel": self.darwinLabel, 
		@"uptimeLabel": self.uptimeLabel,
		@"superview": self.view

	};

	NSString *formatUptimeCenterX = @"V:[superview]-(<=1)-[uptimeLabel]";
	NSString *formatUptimeCenterY = @"H:[superview]-(<=1)-[uptimeLabel]";
	NSString *formatUptimeLeadingTrailing = @"H:|-10-[uptimeLabel]-10-|";

	NSString *formatDarwinBottom = [NSString stringWithFormat:@"V:[darwinLabel]-%d-|", safeInsetBottom];
	NSString *formatDarwinCenterX = @"V:[superview]-(<=1)-[darwinLabel]";
	NSString *formatDarwinLeadingTrailing = @"H:|-10-[darwinLabel]-10-|";

	[self.view addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:formatUptimeCenterX options:NSLayoutFormatAlignAllCenterX metrics:nil views:views]];
	[self.view addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:formatUptimeCenterY options:NSLayoutFormatAlignAllCenterY metrics:nil views:views]];
	[self.view addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:formatUptimeLeadingTrailing options:0 metrics:nil views:views]];

	[self.view addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:formatDarwinBottom options:0 metrics:nil views:views]];
	[self.view addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:formatDarwinCenterX options:NSLayoutFormatAlignAllCenterX metrics:nil views:views]];
	[self.view addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:formatDarwinLeadingTrailing options:0 metrics:nil views:views]];

}


- (void)didTapSettingsButton {

	SettingsVC *settingsVC = [SettingsVC new];
	[self presentViewController:settingsVC animated:YES completion:nil];	

}


- (void)animateLabel:(NSString *)newText withCharacterDelay:(NSTimeInterval)delay {

	// https://stackoverflow.com/questions/11686642/letter-by-letter-animation-for-uilabel

	self.uptimeLabel.text = @"";

	for(int i = 0; i < newText.length; i++) {

		dispatch_async(dispatch_get_main_queue(), ^{

			self.uptimeLabel.text = [NSString stringWithFormat:@"%@%C", self.uptimeLabel.text, [newText characterAtIndex:i]];

		});

		[NSThread sleepForTimeInterval:delay];

	}

}


- (NSString *)setUptimeText {

	[[TaskManager sharedInstance] launchTaskWithArguments:@[@"-c", @"uptime"]];

	NSString *uptimeString = [TaskManager sharedInstance].outputString;

	return uptimeString;

}


// MARK: - NSNotificationCenter

- (void)setDarwinText {

	[[UserDefaultsManager sharedInstance] loadSwitchState];

	BOOL state = [UserDefaultsManager sharedInstance].switchState;

	if(!state) {

		self.darwinLabel.text = nil;
		return;

	}

	[[TaskManager sharedInstance] launchTaskWithArguments:@[@"-c", @"uname -a"]];

	NSString *darwinString = [TaskManager sharedInstance].outputString;

	self.darwinLabel.text = darwinString;

}


- (void)updateAccentColors {

	self.darwinLabel.textColor = [ColorManager sharedInstance].accentColor;
	self.uptimeLabel.textColor = [ColorManager sharedInstance].accentColor;
	self.settingsButton.tintColor = [ColorManager sharedInstance].accentColor;

}


// MARK: - NSTimer

- (void)updateUptimeText {

	self.uptimeLabel.text = [self setUptimeText];

}


@end

File: Cora/Swift/UIKit/Controllers/UPTRootVC.swift
import UIKit


final class UPTRootVC: UIViewController {

	let settingsButton: UIButton = {
		let button = UIButton()
		button.setImage(UIImage(systemName: "gear"), for: .normal)
		return button
	}()

	let darwinLabel: UILabel = {
		let label = UILabel()
		label.font = UIFont(name: "Courier", size: 12)
		label.textAlignment = .center
		label.numberOfLines = 0
		label.translatesAutoresizingMaskIntoConstraints = false
		return label
	}()

	let uptimeLabel: UILabel = {
		let label = UILabel()
		label.font = UIFont(name: "Courier", size: 16)
		label.textAlignment = .center
		label.numberOfLines = 0
		label.translatesAutoresizingMaskIntoConstraints = false
		return label
	}()

	required init?(coder aDecoder: NSCoder) {

		super.init(coder: aDecoder)

	}

	init() {

		super.init(nibName: nil, bundle: nil)

		setupDefaults()
		setupUI()
		setupObservers()

		settingsButton.addTarget(self, action: #selector(didTapSettingsButton), for: .touchUpInside)

		Timer.scheduledTimer(timeInterval: 5, target: self, selector: #selector(updateUptimeText), userInfo: nil, repeats: true)

	}

	private func setupDefaults() {

		UserDefaultsManager.sharedInstance.loadAccentColor()
		UserDefaultsManager.sharedInstance.loadSwitchState()

		guard let switchState = UserDefaultsManager.sharedInstance.switchState else {
			return
		}

		guard switchState else { return }

		setDarwinText()

	}

	private func setupObservers() {

		NotificationCenter.default.removeObserver(self)
		NotificationCenter.default.addObserver(self, selector: #selector(setDarwinText), name: Notification.Name("launchChosenTask"), object: nil)
		NotificationCenter.default.addObserver(self, selector: #selector(updateAccentColors), name: Notification.Name("updateAccentColors"), object: nil)

	}

	override func viewDidLoad() {

		super.viewDidLoad()

		// Do any additional setup after loading the view, typically from a nib.

		view.backgroundColor = UIColor.systemBackground

		UINavigationBar.appearance().shadowImage = UIImage()
		UINavigationBar.appearance().setBackgroundImage(UIImage(), for: .default)

	}

	override func viewDidLayoutSubviews() {

		super.viewDidLayoutSubviews()

		layoutUI()

	}

	private func setupUI() {

		view.addSubview(uptimeLabel)
		view.addSubview(darwinLabel)
		view.addSubview(settingsButton)

		uptimeLabel.text = setUptimeText()
		animateUptimeLabel()

		let settingsButtonItem = UIBarButtonItem(customView: settingsButton)
		navigationItem.rightBarButtonItem = settingsButtonItem

		updateAccentColors()

	}

	private func layoutUI() {

		let safeInsetsBottom = view.safeAreaInsets.bottom

		guard let view = view else {
			return
		}

		let views = [

			"darwinLabel": darwinLabel,
			"uptimeLabel": uptimeLabel,
			"superview": view

		]

		let formatCenterX = "V:[superview]-(<=1)-[uptimeLabel]"
		let formatCenterY = "H:[superview]-(<=1)-[uptimeLabel]"
		let formatLeadingTrailing = "H:|-10-[uptimeLabel]-10-|"

		let darwinFormatBottom = "V:[darwinLabel]-\(safeInsetsBottom)-|"
		let darwinFormatCenterX = "V:[superview]-(<=1)-[darwinLabel]"
		let darwinFormatLeadingTrailing = "H:|-10-[darwinLabel]-10-|"

		let centerXConstraint = NSLayoutConstraint.constraints(withVisualFormat: formatCenterX, options: [.alignAllCenterX], metrics: nil, views: views)
		let centerYConstraint = NSLayoutConstraint.constraints(withVisualFormat: formatCenterY, options: [.alignAllCenterY], metrics: nil, views: views)
		let leadingTrailingConstraint = NSLayoutConstraint.constraints(withVisualFormat: formatLeadingTrailing, options: [], metrics: nil, views: views)

		let darwinBottomConstraint = NSLayoutConstraint.constraints(withVisualFormat: darwinFormatBottom, options: [], metrics: nil, views: views)
		let darwinCenterXConstraint = NSLayoutConstraint.constraints(withVisualFormat: darwinFormatCenterX, options: [.alignAllCenterX], metrics: nil, views: views)
		let darwinLeadingTrailingConstraint = NSLayoutConstraint.constraints(withVisualFormat: darwinFormatLeadingTrailing, options: [], metrics: nil, views: views)

		view.addConstraints(centerXConstraint + centerYConstraint + leadingTrailingConstraint) 
		view.addConstraints(darwinBottomConstraint + darwinCenterXConstraint + darwinLeadingTrailingConstraint)

	}

	@objc private func didTapSettingsButton() {

		let settingsVC = SettingsVC()
		present(settingsVC, animated: true)

	}

	private func animateUptimeLabel() {

		uptimeLabel.text = ""
		let finalText = TaskManager.sharedInstance.outputString ?? ""
		var charIndex = 0.0

		for letter in finalText {
			Timer.scheduledTimer(withTimeInterval: 0.020 * charIndex, repeats: false) { timer in
				self.uptimeLabel.text?.append(letter)
			}
			charIndex += 1
		}

	}

	private func setUptimeText() -> String {

		TaskManager.sharedInstance.launchTask(withArguments: ["-c", "uptime"])

		let uptimeString = TaskManager.sharedInstance.outputString ?? ""

		return uptimeString

	}

	// MARK: - NSNotificationCenter

	@objc private func setDarwinText() {

		UserDefaultsManager.sharedInstance.loadSwitchState()

		guard let switchState = UserDefaultsManager.sharedInstance.switchState else {
			return
		}

		guard switchState else {
			darwinLabel.text = nil
			return 
		}

		TaskManager.sharedInstance.launchTask(withArguments: ["-c", "uname -a"])

		let darwinString = TaskManager.sharedInstance.outputString ?? ""

		darwinLabel.text = darwinString

	}

	@objc private func updateAccentColors() {

 		darwinLabel.textColor = ColorManager.sharedInstance.accentColor
		uptimeLabel.textColor = ColorManager.sharedInstance.accentColor
		settingsButton.tintColor = ColorManager.sharedInstance.accentColor

	}

	// MARK: - NSTimer

	@objc private func updateUptimeText() {

		uptimeLabel.text = setUptimeText()

	}

}
